File: pipeline/rank.py
"""The three rankers: the course tf-idf cosine, log tf-idf cosine and BM25.

Scoring is postings-driven. A query touches only the postings of its own
terms, accumulating into a dict of `doc_idx -> partial score`, so cost is the
length of those posting lists and not the size of the collection. For the two
cosine rankers the per-document norms are computed once when the `Scorer` is
built, in one pass over every posting.

The operation order below is load-bearing: Task 7 ports these three functions
to TypeScript and a parity test compares top-10s document by document. A score
is accumulated per document as the sum, over the *distinct* query terms in
order of first appearance, of that term's contribution; for the cosine rankers
the sum is divided once, at the end, by `norm_d * norm_q`. Weights use
`math.log10`, BM25's idf uses `math.log`, and nothing here uses numpy: plain
Python floats are IEEE doubles, which is what JavaScript numbers are too.
"""

from __future__ import annotations

import math
from bisect import bisect_left

from pipeline.constants import BM25_B, BM25_K1, TOP_N
from pipeline.index import Index
from pipeline.tokenize import analyze
from pipeline.topics import Topic, query_text

RANKERS = {
    "tfidf_raw": "tf-idf cosine (course weighting)",
    "tfidf_log": "log tf-idf cosine",
    "bm25": "BM25",
}

_COSINE = ("tfidf_raw", "tfidf_log")


class Scorer:
    """One ranker bound to one index, with idf and document norms cached.

    Build it once per (index, ranker, k1, b) and reuse it for every query: the
    constructor walks all postings for the cosine rankers, which costs about
    as much as a hundred queries.
    """

    def __init__(
        self, index: Index, ranker: str, k1: float = BM25_K1, b: float = BM25_B
    ) -> None:
        if ranker not in RANKERS:
            raise ValueError(
                f"unknown ranker {ranker!r}; expected one of {list(RANKERS)}"
            )
        self.index = index
        self.ranker = ranker
        self.k1 = k1
        self.b = b
        self._idf: dict[str, float] = {}
        self._norm: list[float] = self._doc_norms() if ranker in _COSINE else []

    # -- weights -----------------------------------------------------------

    def idf(self, term: str) -> float:
        """Inverse document frequency; 0.0 for a term outside the vocabulary.

        `log10(N / df)` for the two cosine rankers, the BM25-plus-one variant
        `ln((N - df + 0.5) / (df + 0.5) + 1)` for BM25, which is never
        negative and so cannot subtract score for a very common term.
        """
        cached = self._idf.get(term)
        if cached is not None:
            return cached
        df = self.index.df(term)
        if df == 0:
            value = 0.0
        elif self.ranker == "bm25":
            n = self.index.n_docs
            value = math.log((n - df + 0.5) / (df + 0.5) + 1)
        else:
            value = math.log10(self.index.n_docs / df)
        self._idf[term] = value
        return value

    def _tf_weight(self, tf: int) -> float:
        """The tf half of a cosine weight, used for documents and queries."""
        if self.ranker == "tfidf_log":
            return 1 + math.log10(tf)
        return float(tf)

    def _doc_norms(self) -> list[float]:
        """Euclidean norm of every document vector, in one pass over postings."""
        norm_sq = [0.0] * self.index.n_docs
        for term, plist in self.index.postings.items():
            idf = self.idf(term)
            if idf == 0.0:
                continue
            for doc_idx, tf in plist:
                w = self._tf_weight(tf) * idf
                norm_sq[doc_idx] += w * w
        return [math.sqrt(v) for v in norm_sq]

    def _query_tf(self, query_terms: list[str]) -> dict[str, int]:
        """Distinct query terms that are in the vocabulary, in query order,
        mapped to their query term frequency. Dicts keep insertion order, so
        this fixes the accumulation order the TypeScript port must mirror."""
        qtf: dict[str, int] = {}
        for term in query_terms:
            if term in qtf:
                qtf[term] += 1
            elif term in self.index.postings:
                qtf[term] = 1
        return qtf

    # -- scoring -----------------------------------------------------------

    def score(self, query_terms: list[str]) -> dict[int, float]:
        """`doc_idx -> score` for every document that scores above zero."""
        qtf = self._query_tf(query_terms)
        if not qtf:
            return {}
        if self.ranker == "bm25":
            return self._score_bm25(qtf)
        return self._score_cosine(qtf)

    def _score_cosine(self, qtf: dict[str, int]) -> dict[int, float]:
        postings = self.index.postings
        dot: dict[int, float] = {}
        norm_q_sq = 0.0
        for term, q in qtf.items():
            idf = self.idf(term)
            w_q = self._tf_weight(q) * idf
            norm_q_sq += w_q * w_q
            if w_q == 0.0:
                continue
            for doc_idx, tf in postings[term]:
                w_d = self._tf_weight(tf) * idf
                dot[doc_idx] = dot.get(doc_idx, 0.0) + w_q * w_d
        norm_q = math.sqrt(norm_q_sq)
        if norm_q == 0.0:
            return {}
        norm = self._norm
        scores = {}
        for doc_idx, value in dot.items():
            norm_d = norm[doc_idx]
            if norm_d == 0.0 or value <= 0.0:
                continue
            scores[doc_idx] = value / (norm_d * norm_q)
        return scores

    def _score_bm25(self, qtf: dict[str, int]) -> dict[int, float]:
        postings = self.index.postings
        doc_len = self.index.doc_len
        avgdl = self.index.avgdl
        k1, b = self.k1, self.b
        scores: dict[int, float] = {}
        for term, q in qtf.items():
            idf = self.idf(term)
            if idf == 0.0:
                continue
            for doc_idx, tf in postings[term]:
                dl = doc_len[doc_idx]
                contribution = (
                    q * idf * tf * (k1 + 1) / (tf + k1 * (1 - b + b * dl / avgdl))
                )
                scores[doc_idx] = scores.get(doc_idx, 0.0) + contribution
        return {d: s for d, s in scores.items() if s > 0.0}

    def rank(
        self, query_terms: list[str], top_n: int = TOP_N
    ) -> list[tuple[int, float]]:
        """The top `top_n` documents, highest score first, ties by doc index."""
        scores = self.score(query_terms)
        ranked = sorted(scores.items(), key=lambda item: (-item[1], item[0]))
        return ranked[:top_n]

    # -- explanation -------------------------------------------------------

    def _tf(self, term: str, doc_idx: int) -> int:
        """Term frequency of `term` in one document, 0 when it does not occur."""
        plist = self.index.postings.get(term)
        if not plist:
            return 0
        pos = bisect_left(plist, doc_idx, key=lambda posting: posting[0])
        if pos < len(plist) and plist[pos][0] == doc_idx:
            return plist[pos][1]
        return 0

    def explain(self, query_terms: list[str], doc_idx: int) -> list[dict]:
        """One row per distinct query term that the document actually has.

        Rows are in query order and their `contribution` values sum to
        `score(query_terms)[doc_idx]`, so the site can show the arithmetic of
        a result without the reader having to trust it. Terms outside the
        vocabulary, and terms absent from this document, are left out.
        """
        qtf = self._query_tf(query_terms)
        if not qtf:
            return []
        rows: list[dict] = []
        if self.ranker == "bm25":
            k1, b = self.k1, self.b
            dl = self.index.doc_len[doc_idx]
            avgdl = self.index.avgdl
            for term, q in qtf.items():
                tf = self._tf(term, doc_idx)
                if tf == 0:
                    continue
                idf = self.idf(term)
                rows.append(
                    {
                        "term": term,
                        "qtf": q,
                        "tf": tf,
                        "df": self.index.df(term),
                        "idf": idf,
                        "contribution": q
                        * idf
                        * tf
                        * (k1 + 1)
                        / (tf + k1 * (1 - b + b * dl / avgdl)),
                    }
                )
            return rows

        norm_q = math.sqrt(
            sum((self._tf_weight(q) * self.idf(term)) ** 2 for term, q in qtf.items())
        )
        denom = self._norm[doc_idx] * norm_q
        if denom == 0.0:
            # Every query term is in every document, so no document scores at
            # all and there is no arithmetic to show.
            return []
        for term, q in qtf.items():
            tf = self._tf(term, doc_idx)
            if tf == 0:
                continue
            idf = self.idf(term)
            w_q = self._tf_weight(q) * idf
            w_d = self._tf_weight(tf) * idf
            rows.append(
                {
                    "term": term,
                    "qtf": q,
                    "tf": tf,
                    "df": self.index.df(term),
                    "idf": idf,
                    "contribution": w_q * w_d / denom,
                }
            )
        return rows


def run_topics(
    index: Index,
    scorer: Scorer,
    topics: list[Topic],
    field: str,
    stopwords: frozenset[str] | None,
    top_n: int = TOP_N,
) -> dict[int, list[tuple[str, float]]]:
    """One run: `topic number -> [(docno, score)]`, ranked.

    The query is analysed exactly as the index was built, which is the whole
    point of passing the index in: a stemmed index gets stemmed query terms
    and an index built with stopwords removed gets them removed from the
    query too.
    """
    stops = stopwords if index.stop else None
    docnos = index.docnos
    runs: dict[int, list[tuple[str, float]]] = {}
    for topic in topics:
        terms = analyze(query_text(topic, field), stops, index.stem)
        runs[topic.num] = [
            (docnos[doc_idx], score) for doc_idx, score in scorer.rank(terms, top_n)
        ]
    return runs

File: site/src/search/rank.ts
/**
 * The two rankers the page offers, ported from `pipeline/rank.py` (W3).
 *
 * Scoring is postings-driven: a query touches only the posting lists of its
 * own terms and accumulates into a `Float64Array` keyed by document index, so
 * cost is the length of those lists and not the size of the collection.
 *
 * The operation order below is load-bearing. A score is accumulated per
 * document as the sum, over the *distinct* query terms in order of first
 * appearance, of that term's contribution; for the cosine ranker the sum is
 * divided once, at the end, by `norm_d * norm_q`. Cosine weights use
 * `Math.log10`, BM25's idf uses `Math.log`. Floating-point addition is not
 * associative, and `tests/parity.test.ts` compares top tens document by
 * document against the Python run, so none of this is free to be tidied.
 */

import type { SearchIndex } from "./index.ts";
import { analyze } from "./tokenize.ts";

export type Ranker = "bm25" | "tfidf_raw";

export const RANKER_LABELS: Record<Ranker, string> = {
  bm25: "BM25",
  tfidf_raw: "tf-idf cosine (course weighting)",
};

export function isRanker(value: string): value is Ranker {
  return value === "bm25" || value === "tfidf_raw";
}

/** One row of the arithmetic behind a single result. */
export interface ExplainRow {
  term: string;
  qtf: number;
  tf: number;
  df: number;
  idf: number;
  contribution: number;
}

export interface Hit {
  docIdx: number;
  no: string;
  date: string;
  headline: string;
  score: number;
  explain: ExplainRow[];
}

export interface SearchResult {
  hits: Hit[];
  /** Distinct query stems, in order of first appearance. */
  terms: string[];
  /** The subset of `terms` with df 0: stems this corpus has never seen. */
  unknown: string[];
  /** How many articles scored above zero, of which `hits` is the top slice. */
  total: number;
}

/**
 * Document norms are an index-wide property of one ranker, so they are built
 * once and reused. A `WeakMap` keeps them out of the `SearchIndex` shape and
 * lets the index be collected with them.
 */
const NORM_CACHE = new WeakMap<SearchIndex, Map<Ranker, Float64Array>>();

/**
 * Inverse document frequency; 0 for a term outside the vocabulary.
 *
 * `log10(N / df)` for the cosine ranker, the BM25-plus-one variant
 * `ln((N - df + 0.5) / (df + 0.5) + 1)` for BM25, which is never negative and
 * so cannot subtract score for a very common term. Note that the cosine idf
 * is exactly 0 when a term is in every document, which drops that term from
 * the query and from every document vector.
 */
function idfOf(index: SearchIndex, df: number, ranker: Ranker): number {
  if (df === 0) return 0;
  if (ranker === "bm25") {
    return Math.log((index.nDocs - df + 0.5) / (df + 0.5) + 1);
  }
  return Math.log10(index.nDocs / df);
}

function idfOfTerm(index: SearchIndex, term: string, ranker: Ranker): number {
  const postings = index.terms.get(term);
  return idfOf(index, postings ? postings.docs.length : 0, ranker);
}

/** Euclidean norm of every document vector, in one pass over the postings. */
function docNorms(index: SearchIndex, ranker: Ranker): Float64Array {
  let perRanker = NORM_CACHE.get(index);
  if (!perRanker) {
    perRanker = new Map();
    NORM_CACHE.set(index, perRanker);
  }
  const cached = perRanker.get(ranker);
  if (cached) return cached;

  const normSq = new Float64Array(index.nDocs);
  for (const postings of index.terms.values()) {
    const idf = idfOf(index, postings.docs.length, ranker);
    if (idf === 0) continue;
    for (let i = 0; i < postings.docs.length; i += 1) {
      const w = postings.tf[i] * idf;
      normSq[postings.docs[i]] += w * w;
    }
  }
  for (let d = 0; d < normSq.length; d += 1) normSq[d] = Math.sqrt(normSq[d]);
  perRanker.set(ranker, normSq);
  return normSq;
}

/**
 * Distinct query terms that are in the vocabulary, in query order, mapped to
 * their query term frequency. `Map` keeps insertion order, which fixes the
 * accumulation order the Python dict fixes on the other side.
 */
function queryTf(index: SearchIndex, stems: string[]): Map<string, number> {
  const qtf = new Map<string, number>();
  for (const term of stems) {
    const seen = qtf.get(term);
    if (seen !== undefined) qtf.set(term, seen + 1);
    else if (index.terms.has(term)) qtf.set(term, 1);
  }
  return qtf;
}

/** Term frequency of `term` in one document; 0 when it does not occur. */
function tfIn(index: SearchIndex, term: string, docIdx: number): number {
  const postings = index.terms.get(term);
  if (!postings) return 0;
  let lo = 0;
  let hi = postings.docs.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    const d = postings.docs[mid];
    if (d === docIdx) return postings.tf[mid];
    if (d < docIdx) lo = mid + 1;
    else hi = mid - 1;
  }
  return 0;
}

interface Accumulator {
  score: Float64Array;
  touched: number[];
}

function scoreBm25(
  index: SearchIndex,
  qtf: Map<string, number>,
): Accumulator | null {
  const { k1, b, avgdl, docLen } = index;
  const score = new Float64Array(index.nDocs);
  const seen = new Uint8Array(index.nDocs);
  const touched: number[] = [];
  for (const [term, q] of qtf) {
    const postings = index.terms.get(term);
    if (!postings) continue;
    const idf = idfOf(index, postings.docs.length, "bm25");
    if (idf === 0) continue;
    for (let i = 0; i < postings.docs.length; i += 1) {
      const d = postings.docs[i];
      const tf = postings.tf[i];
      const dl = docLen[d];
      score[d] += (q * idf * tf * (k1 + 1)) / (tf + k1 * (1 - b + (b * dl) / avgdl));
      if (seen[d] === 0) {
        seen[d] = 1;
        touched.push(d);
      }
    }
  }
  return { score, touched };
}

function scoreCosine(
  index: SearchIndex,
  qtf: Map<string, number>,
): Accumulator | null {
  const dot = new Float64Array(index.nDocs);
  const seen = new Uint8Array(index.nDocs);
  const touched: number[] = [];
  let normQSq = 0;
  for (const [term, q] of qtf) {
    const postings = index.terms.get(term);
    if (!postings) continue;
    const idf = idfOf(index, postings.docs.length, "tfidf_raw");
    const wq = q * idf;
    normQSq += wq * wq;
    if (wq === 0) continue;
    for (let i = 0; i < postings.docs.length; i += 1) {
      const d = postings.docs[i];
      dot[d] += wq * (postings.tf[i] * idf);
      if (seen[d] === 0) {
        seen[d] = 1;
        touched.push(d);
      }
    }
  }
  const normQ = Math.sqrt(normQSq);
  if (normQ === 0) return null;

  const norm = docNorms(index, "tfidf_raw");
  const score = dot;
  const kept: number[] = [];
  for (const d of touched) {
    const normD = norm[d];
    if (normD === 0 || score[d] <= 0) {
      score[d] = 0;
      continue;
    }
    score[d] = score[d] / (normD * normQ);
    kept.push(d);
  }
  return { score, touched: kept };
}

/** The query's own norm, in the same term order the score used. */
function queryNorm(
  index: SearchIndex,
  qtf: Map<string, number>,
  ranker: Ranker,
): number {
  let sum = 0;
  for (const [term, q] of qtf) {
    const w = q * idfOfTerm(index, term, ranker);
    sum += w * w;
  }
  return Math.sqrt(sum);
}

/**
 * One row per distinct query term the document actually has, in query order,
 * whose contributions sum to the document's score. Terms outside the
 * vocabulary, and terms absent from this document, are left out.
 */
function explainHit(
  index: SearchIndex,
  qtf: Map<string, number>,
  ranker: Ranker,
  docIdx: number,
): ExplainRow[] {
  const rows: ExplainRow[] = [];
  if (ranker === "bm25") {
    const { k1, b, avgdl } = index;
    const dl = index.docLen[docIdx];
    for (const [term, q] of qtf) {
      const tf = tfIn(index, term, docIdx);
      if (tf === 0) continue;
      const df = index.terms.get(term)?.docs.length ?? 0;
      const idf = idfOf(index, df, "bm25");
      rows.push({
        term,
        qtf: q,
        tf,
        df,
        idf,
        contribution:
          (q * idf * tf * (k1 + 1)) / (tf + k1 * (1 - b + (b * dl) / avgdl)),
      });
    }
    return rows;
  }

  const denom = docNorms(index, ranker)[docIdx] * queryNorm(index, qtf, ranker);
  // Every query term is in every document, so nothing scores at all and there
  // is no arithmetic to show.
  if (denom === 0) return rows;
  for (const [term, q] of qtf) {
    const tf = tfIn(index, term, docIdx);
    if (tf === 0) continue;
    const df = index.terms.get(term)?.docs.length ?? 0;
    const idf = idfOf(index, df, ranker);
    const wq = q * idf;
    const wd = tf * idf;
    rows.push({ term, qtf: q, tf, df, idf, contribution: (wq * wd) / denom });
  }
  return rows;
}

/**
 * Rank the collection for one query string.
 *
 * The query is analysed exactly as the index was built -- stopwords removed,
 * then stemmed -- because the index it is searching was built that way.
 * Ties are broken by ascending document index, after descending score, which
 * is what `sorted(..., key=lambda item: (-item[1], item[0]))` does in Python.
 */
export function search(
  index: SearchIndex,
  query: string,
  ranker: Ranker,
  k = 10,
): SearchResult {
  const stems = analyze(query, true, true);
  const terms: string[] = [];
  const seen = new Set<string>();
  for (const term of stems) {
    if (seen.has(term)) continue;
    seen.add(term);
    terms.push(term);
  }
  const unknown = terms.filter((term) => !index.terms.has(term));

  const qtf = queryTf(index, stems);
  if (qtf.size === 0) return { hits: [], terms, unknown, total: 0 };

  const acc =
    ranker === "bm25" ? scoreBm25(index, qtf) : scoreCosine(index, qtf);
  if (!acc) return { hits: [], terms, unknown, total: 0 };

  const scored: number[] = [];
  for (const d of acc.touched) if (acc.score[d] > 0) scored.push(d);
  scored.sort((a, bIdx) => acc.score[bIdx] - acc.score[a] || a - bIdx);

  const hits = scored.slice(0, k).map((docIdx) => {
    const doc = index.docs[docIdx];
    return {
      docIdx,
      no: doc.no,
      date: doc.d,
      headline: doc.h,
      score: acc.score[docIdx],
      explain: explainHit(index, qtf, ranker, docIdx),
    };
  });

  return { hits, terms, unknown, total: scored.length };
}
